src/backend/mpi/swdsm.cpp

```diff
@@ -594,346 +512,319 @@ unsigned long getOffset(unsigned long addr){
+	if(tag != GLOBAL_NULL && idx != GLOBAL_NULL && cacheControl[idx].dirty == DIRTY){
+		mprotect((char*)startAddr+tag,CACHELINE*pagesize,PROT_READ);
+		for(i = 0; i <CACHELINE; i++){
+			storepageDIFF(idx+i,tag+pagesize*i);
+			cacheControl[idx+i].dirty = CLEAN;
 		}
-		for(i = 0; i < (unsigned long)numtasks; i++){
-			if(barwindowsused[i] == 1){
-				MPI_Win_unlock(i, globalDataWindow[i]);
-				barwindowsused[i] = 0;
-			}
+	}
+	for(i = 0; i < (unsigned long)numtasks; i++){
+		if(barwindowsused[i] == 1){
+			MPI_Win_unlock(i, globalDataWindow[i]);
+			barwindowsused[i] = 0;
 		}
-		writebufferstart = (writebufferstart+1)%writebuffersize;
-		sem_post(&ibsem);
-		sem_post(&writerwaitsem);
 	}
-	return nullptr;
+	writebufferstart = (writebufferstart+1)%writebuffersize;
+	sem_post(&ibsem);
 }
 
-void * loadcacheline(void * x){
-	UNUSED_PARAM(x);
+void load_cache_entry(unsigned long loadtag, unsigned long loadline) {
 	int i;
 	unsigned long homenode;
 	unsigned long id = 1 << getID();
 	unsigned long invid = ~id;
-	while(1){
-		sem_wait(&loadstartsem);
-		if(loadtag[0]>=size_of_all){//Trying to access/prefetch out of memory
-			sem_post(&loadwaitsem);
-			continue;
-		}
-		homenode = getHomenode(loadtag[0]);
-		unsigned long cacheIndex = loadline[0];
-		if(cacheIndex >= cachesize){
-			printf("idx > size   cacheIndex:%ld cachesize:%ld\n",cacheIndex,cachesize);
-			sem_post(&loadwaitsem);
-			continue;
-		}
-		sem_wait(&ibsem);
+
+	if(loadtag>=size_of_all){//Trying to access/prefetch out of memory
+		return;
+	}
+	homenode = getHomenode(loadtag);
+	unsigned long cacheIndex = loadline;
+	if(cacheIndex >= cachesize){
+		printf("idx > size   cacheIndex:%ld cachesize:%ld\n",cacheIndex,cachesize);
+		return;
+	}
+	sem_wait(&ibsem);
 
 
-		unsigned long pageAddr = loadtag[0];
-		unsigned long blocksize = pagesize*CACHELINE;
-		unsigned long lineAddr = pageAddr/blocksize;
-		lineAddr *= blocksize;
+	unsigned long pageAddr = loadtag;
+	unsigned long blocksize = pagesize*CACHELINE;
+	unsigned long lineAddr = pageAddr/blocksize;
+	lineAddr *= blocksize;
 
-		unsigned long startidx = cacheIndex/CACHELINE;
-		startidx*=CACHELINE;
-		unsigned long end = startidx+CACHELINE;
+	unsigned long startidx = cacheIndex/CACHELINE;
+	startidx*=CACHELINE;
+	unsigned long end = startidx+CACHELINE;
 
-		if(end>=cachesize){
-			end = cachesize;
-		}
+	if(end>=cachesize){
+		end = cachesize;
+	}
 
-		argo_byte tmpstate = cacheControl[startidx].state;
-		unsigned long tmptag = cacheControl[startidx].tag;
+	argo_byte tmpstate = cacheControl[startidx].state;
+	unsigned long tmptag = cacheControl[startidx].tag;
 
-		if(tmptag == lineAddr && tmpstate != INVALID){
-			sem_post(&ibsem);
-			sem_post(&loadwaitsem);
-			continue;
-		}
+	if(tmptag == lineAddr && tmpstate != INVALID){
+		sem_post(&ibsem);
+		return;
+	}
 
 
-		void * lineptr = (char*)startAddr + lineAddr;
+	void * lineptr = (char*)startAddr + lineAddr;
 
+	if(cacheControl[startidx].tag  != lineAddr){
```

Comment:
What's happening here? Why are there two if statements, one after the other, with the same condition?

Reply:
If you want historical explanations please ask @magnusandersnorgren - I'd assume remnants of debugging.
Otherwise I'll interpret this as "please remove superfluous if()".

Reply:
Let's go with "please remove superfluous if()"

Reply:
I will leave it up to you if you want to do this in this PR or leave it for later. Since it's not a functional change I am going to go ahead and not care.

Reply:
This will be resolved together with #32, as the entire code will probably change.